src/Storage/Device.php

```diff
@@ -38,34 +38,51 @@ abstract public function getRoot(): string;
      *
      * Each device hold a complex directory structure that is being build in this method.
      *
-     * @param $filename
+     * @param string $filename
+     * @param string $prefix
      *
      * @return string
      */
-    abstract public function getPath(string $filename): string;
+    abstract public function getPath(string $filename, string $prefix = null): string;
 
     /**
      * Upload.
      *
-     * Upload a file to desired destination in the selected disk, return true on success and false on failure.
+     * Upload a file to desired destination in the selected disk
+     * return number of chunks uploaded or 0 if it fails.
      *
      * @param string $source
      * @param string $path
-     *
+     * @param int $chunk
+     * @param int $chunks
+     * @param array $metadata
+     * 
      * @throws \Exception
      *
+     * @return int
+     */
+    abstract public function upload(string $source, string $path, int $chunk = 1, int $chunks = 1, array &$metadata = []): int;
+
+    /**
+     * Abort Chunked Upload
+     * 
+     * @param string $path
+     * @param mixed $extra
+     * 
      * @return bool
      */
-    abstract public function upload($source, $path): bool;
+    abstract public function abort(string $path, mixed $extra = ''): bool;
```

Comment:
What is extra, what is it mixed and if it is mixed why the default is string?

Reply:
To abort chunk file, s3 requires uploadId, that's the extra param. We don't need this for local adapter. Made it string, but now sure what would be a good name for it.